core/store/src/trie/shard_tries.rs

```diff
@@ -69,19 +31,51 @@ struct ShardTriesInner {
 pub struct ShardTries(Arc<ShardTriesInner>);
 
 impl ShardTries {
-    pub fn new(store: Store, trie_cache_factory: TrieCacheFactory) -> Self {
-        let caches = trie_cache_factory.create_initial_caches(false);
-        let view_caches = trie_cache_factory.create_initial_caches(true);
+    pub fn new(store: Store, trie_config: TrieConfig, initial_shard_layout: &ShardLayout) -> Self {
+        let shard_uids = initial_shard_layout.get_shard_uids();
+        let caches = Self::create_initial_caches(&trie_config, &shard_uids, false);
+        let view_caches = Self::create_initial_caches(&trie_config, &shard_uids, true);
         ShardTries(Arc::new(ShardTriesInner {
             store,
-            trie_cache_factory,
+            trie_config,
             caches: RwLock::new(caches),
             view_caches: RwLock::new(view_caches),
         }))
     }
 
+    /// Create `ShardTries` with a fixed number of shards with shard version 0.
+    ///
+    /// If you test cares about the shard version, use `test_shard_version` instead.
     pub fn test(store: Store, num_shards: NumShards) -> Self {
-        Self::new(store, TrieCacheFactory::new(Default::default(), 0, num_shards))
+        let version = 0;
+        Self::test_shard_version(store, version, num_shards)
+    }
+
+    pub fn test_shard_version(store: Store, version: ShardVersion, num_shards: NumShards) -> Self {
```

Comment:
Maybe we should pass shard_uids to `new` so that we can unfork the two methods? Even if don't change the API here, let's introduce a common helped ctor which can be used by both `new` and `test_shard_version`. 

It's a smell where the different constructors don't converge to a single, root one.

Reply:
passing `shard_uids` to `new` seems better to me, I've done that now